From 74dba3599fd7fa042b253dce1a6742d553c0e54c Mon Sep 17 00:00:00 2001
From: Jim Stichnoth <jstichnoth@mythtv.org>
Date: Thu, 23 Apr 2015 17:35:29 -0700
Subject: [PATCH] Subtitles: Fix state calculation in
 EnableCaptions/DisableCaptions.

26f6437b874d84c87a52df6d33e449c0d5d9a4bc was too aggressive in
assuming that EnableCaptions would actually enable text subtitles, and
DisableCaptions would actually disable text subtitles.  They only
enable/disable a particular type of subtitles (which may have already
been enabled/disabled), so we need to check the overall state after
the operation is performed.

(cherry picked from commit 0925cc05806a057b6549a3471ee043d68948d0bd)
---
 mythtv/libs/libmythtv/mythplayer.cpp |    4 ++--
 mythtv/libs/libmythtv/mythplayer.h   |    3 +++
 2 files changed, 5 insertions(+), 2 deletions(-)

diff --git a/mythtv/libs/libmythtv/mythplayer.cpp b/mythtv/libs/libmythtv/mythplayer.cpp
index a62ab31..59ebd1b 100644
--- a/mythtv/libs/libmythtv/mythplayer.cpp
+++ b/mythtv/libs/libmythtv/mythplayer.cpp
@@ -1377,7 +1377,7 @@ void MythPlayer::DisableCaptions(uint mode, bool osd_msg)
 
     QMutexLocker locker(&osdLock);
 
-    textDesired = false;
+    textDesired = textDisplayMode & kDisplayAllTextCaptions;
     QString msg = "";
     if (kDisplayNUVTeletextCaptions & mode)
         msg += tr("TXT CAP");
@@ -1414,7 +1414,7 @@ void MythPlayer::DisableCaptions(uint mode, bool osd_msg)
 void MythPlayer::EnableCaptions(uint mode, bool osd_msg)
 {
     QMutexLocker locker(&osdLock);
-    textDesired = true;
+    textDesired = mode & kDisplayAllTextCaptions;
     QString msg = "";
     if ((kDisplayCC608 & mode) || (kDisplayCC708 & mode) ||
         (kDisplayAVSubtitle & mode) || kDisplayRawTextSubtitle & mode)
diff --git a/mythtv/libs/libmythtv/mythplayer.h b/mythtv/libs/libmythtv/mythplayer.h
index 541a0d5..3ce2da2 100644
--- a/mythtv/libs/libmythtv/mythplayer.h
+++ b/mythtv/libs/libmythtv/mythplayer.h
@@ -77,6 +77,9 @@ enum
     kDisplayRawTextSubtitle     = 0x080,
     kDisplayAllCaptions         = 0x0FF,
     kDisplayTeletextMenu        = 0x100,
+    kDisplayAllTextCaptions     = ~kDisplayDVDButton &
+                                  ~kDisplayAVSubtitle &
+                                   kDisplayAllCaptions,
 };
 
 enum PlayerFlags
-- 
1.7.10.2

